seed/seedwriter/writer.go

```diff
@@ -865,7 +876,95 @@ func (w *Writer) checkBase(info *snap.Info, modes []string) error {
 	return w.policy.checkBase(info, modes, w.availableByMode)
 }
 
-func (w *Writer) downloaded(seedSnaps []*SeedSnap) error {
+func isKernelSnap(sn *SeedSnap) bool {
+	return sn.modelSnap != nil && sn.modelSnap.SnapType == "kernel"
+}
+
+// snapdCarryingSnapsKnown returns true once it all the snaps that carry snapd
+// or parts of it have been captured; they need to be known before assertions
+// can be fetched with the correct max formats
+func (w *Writer) snapdCarryingSnapsKnown() bool {
+	return w.systemSnap != nil && (w.noKernelSnap || w.kernelSnap != nil)
+}
+
+func (w *Writer) considerForSnapdCarrying(sn *SeedSnap) error {
+	if w.systemSnap == nil {
+		if w.policy.isSystemSnapCandidate(sn) {
+			w.systemSnap = sn
+			return nil
+		}
+	}
+	if w.noKernelSnap {
+		return nil
+	}
+	if isKernelSnap(sn) {
+		w.kernelSnap = sn
+		return nil
+	}
+	w.noKernelSnap = true
+	return nil
+}
+
+// An AssertsFetchFunc should fetch appropriate assertions for the snap sn, it
+// can take into account format constraints caused by the given systemSnap and
+// kernelSnap if set. The returned references are expected to be resolvable
+// in the writing assertion database.
+type AssertsFetchFunc func(sn, systemsSnap, kernelSnap *SeedSnap) ([]*asserts.Ref, error)
+
+func (w *Writer) ensureARefs(upToSnap *SeedSnap, fetchAsserts AssertsFetchFunc) error {
+	n := len(w.snapsFromModel)
+	xn := len(w.extraSnaps)
+
+	iterateUpTo := func(count *int, f func(sn *SeedSnap) error) error {
+		i := *count
+		snaps := w.snapsFromModel
+		for ; i < n+xn; i++ {
+			j := i
+			if j >= n {
+				snaps = w.extraSnaps
+				j -= n
+			}
+			sn := snaps[j]
+			if err := f(sn); err != nil {
+				return err
+			}
+			*count += 1
+			if sn == upToSnap {
+				break
+			}
+		}
+		return nil
+	}
+
+	if !w.snapdCarryingSnapsKnown() {
+		iterateUpTo(&w.consideredForSnapdCarrying, w.considerForSnapdCarrying)
```

Comment:
Missing error handling here?

Reply:
w.considerForSnapdCarrying never errors, but I can try to make it more explicit